src/util/psbt/responsibilities.rs

```diff
@@ -0,0 +1,142 @@
+/// [BIP-174](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#creator). As noted
+/// there the both functionalities are most often combined as the creator most likely has more meta
+/// information about the transaction they are creating and is thus able to put it into the PSBT
+/// (which otherwise would be a separate update step).
+pub trait PsbtWallet {
+    /// Wallet backend specific errors like database connection errors that aren't captured in
+    /// `PsbtCreationError`.
+    type WalletError: Debug;
+
+    /// Returns a script to which possible change should be sent
+    fn get_change_script(&mut self) -> Result<Script, Self::WalletError>;
+
+    /// Create a transaction that pays a set of outputs, optionally sending change back to a
+    /// specified change script. If no change script is provided the wallet may overpay as it sees
+    /// fit. This can be used to increase privacy, but should be externally limited for safety
+    /// reasons.
+    fn create_transaction(
+        &mut self,
+        outputs: &[TxOut],
+        change_output: Option<Script>,
+    ) -> Result<PartiallySignedTransaction, PsbtCreationError<Self::WalletError>>;
+
+    /// Creates a transaction that pays a certain `amount` to a `script` and keeps the change
+    fn pay_to_script(
+        &mut self,
+        script_pk: Script,
+        amount: Amount,
+    ) -> Result<PartiallySignedTransaction, PsbtCreationError<Self::WalletError>> {
+        let change_script = self
+            .get_change_script()
+            .map_err(|e| PsbtCreationError::WalletError(e))?;
+
+        self.create_transaction(
+            &[TxOut {
+                value: amount.as_sat(),
+                script_pubkey: script_pk,
+            }],
+            Some(change_script),
+        )
+    }
+
+    /// Creates a transaction that pays a certain `amount` to an `address` and keeps the change
+    fn pay_to_address(
+        &mut self,
+        address: Address,
+        amount: Amount,
+    ) -> Result<PartiallySignedTransaction, PsbtCreationError<Self::WalletError>> {
+        self.pay_to_script(address.script_pubkey(), amount)
+    }
+}
+
+/// Common errors when creating a transaction
+#[derive(Debug)]
+pub enum PsbtCreationError<E: Debug> {
+    /// The wallet doens't control a sufficient amount of Bitcoins to fund the transaction
+    InsufficientFunds,
+    /// One of the outputs has a value below the dust limit
+    OutputBelowDustLimit,
+    /// Too many outputs were supplied
+    TooManyOutputs,
+    /// Wallet backend error
+    WalletError(E),
+}
+
+/// A signer capable of signing PSBTs. This can either be a software signer (see the implementation
+/// for `ExtendedSecretKey`) or a hardware device.
+pub trait SignPsbt {
+    /// Signing backend error type
+    type SignError: Debug;
+
+    /// PSBT validation function that can be used by custom signers too
+    fn validate(psbt: &PartiallySignedTransaction) -> Result<(), PsbtSignError<Self::SignError>> {
```

Comment:
It's quite strange that `validate` function does not take `self`, while `create_transaction` in `PsbtWallet` trait above does...

Reply:
`PsbtSignError<Self::SignError>`: validation can't return signing error, so that's why in my version I had two possible error types per signing trait

Reply:
Yeah, I'm not really happy with the validation either. Do you know if there are any cases where verification needs external data (that's why it currently doesn't take `&self` and I even though about taking it out of the trait making it an independent function)? I don't have a strong opinion on it.

Regarding the error: I don't see a big benefit in a separate error type as every validation error is also a signing error (validation _has_ to happen before signing). But of course we could split them up.

Reply:
The whole rust concept of `Result` and `Error` types is to have a error types that are specific to the procedure. From the lay point and easiness of development, many Error types are just "bulk" errors "cover'em all" style. However, with `match`, when you do `match let Err(err) = some_fn()`, you need to have exhaustive enumeration of possible errors, and mixing validation errors with all possible non-validation errors for a `validate` function makes not a lot of sense and simply not handy. So I prefer to have a separate `ValidationError` and `SigningError` types, with signing function returning "union" of them two

Reply:
Sure, no problem with that, will change it. The reasoning was that in practice probably nobody is going to call `validate` on it's own, so exhaustive matching isn't required if all we do is pass it along as a signing error in the next step.

Reply:
fixed